Improvements to the Step by step tutorial (#6622)
Particular/docs.particular.net: tutorials/nservicebus-step-by-step/3-multiple-endpoints/tutorial.md

## Sending messages

> We've already shown how an endpoint can "send a message to itself" using the `SendLocal()` method, which is available via the `IEndpointInstance` that we used in the endpoint to create a UI, and also via the `IMessageHandlerContext` that can be access while handling a message.

Reviewer: ```suggestion
We've already shown how an endpoint can "send a message to itself" using the `SendLocal()` method, which is available via the `IEndpointInstance` that we used in the endpoint to create a UI, and also via the `IMessageHandlerContext` that can be accessed while handling a message.
```

This concept is called **logical routing**, the mapping of specific message types to logical endpoints that can process those messages. Each command message should have one logical endpoint that owns that message and can process it.

We say *logical routing* because this is at a logical layer only, which isn't necessarily the same as *physical routing*. Within one *logical* endpoint, there may be many *physical* endpoint instances deployed to multiple servers.

> [!NOTE]
> An [**endpoint**](/nservicebus/concepts/glossary.md#endpoint) is a logical concept, defined by an endpoint name and associated implementation, that defines an owner responsible for processing messages.
>
> An [**endpoint instance**](/nservicebus/concepts/glossary.md#endpoint-instance) is a physical instance of the endpoint deployed to a single server. Many endpoint instances may be deployed to many servers in order to scale out the processing of a high-volume message to multiple servers.

For now, we'll only concern ourselves with logical routing, and leave the rest of it (physical routing, scale-out, etc.) for a later time.

> Because logical routing does not cover physical concerns, but only defines logical ownership, this is something that we (as the implementators) should control, and is not an Operations (as in the part of our organization that handles the underlaying hardware infrastructure) concern. While Operations may want to be able to move an endpoint to a different server using only configuration files, changing the owner for messages would require code changes and a recompile/redeploy anyway.

Reviewer: ```suggestion
Because logical routing does not cover physical concerns, but only defines logical ownership, this is something that the code implementation (as opposed to hardware configuration) should control. For example, while staff who work with hardware operations may want to be able to move an endpoint to a different server using only configuration files, changing the owner for messages would require code changes and a recompile/redeploy.
```

However, it's common in NServiceBus solutions to run multiple projects (i.e. endpoints) at once. To make this easier, configure both endpoints (**ClientUI** and **Sales**) to run at startup using Visual Studio's [multiple startup projects](https://msdn.microsoft.com/en-us/library/ms165413.aspx) feature.

If you run the project now, **ClientUI** will work just as it did before, and **Sales** will start up and wait for messages that will never arrive.

### Moving the handler

Now let's move the handler from **ClientUI** over to **Sales** where it belongs.

 1. In the Solution Explorer, find **PlaceOrderHandler.cs** in the **ClientUI** project, and drag it to the **Sales** project.
 1. Open the new **PlaceOrderHandler.cs** in **Sales** and change the namespace from `ClientUI` to `Sales` to match its new home.
 1. Visual Studio's default action when you drag files between projects is to copy them, so you must delete the old **PlaceOrderHandler.cs** from the **ClientUI** endpoint.

> Now that the handler is in the correct endpoint, what would happen if we started the solution? **Sales** now has a message handler, but recall that ClientUI is still calling `endpointInstance.SendLocal(command)` which effectively sends the message to itself. But it doesn't have a handler for this message anymore.

Reviewer: ```suggestion
Now that the handler is in the correct endpoint, what would happen if we started the solution? **Sales** now has a message handler, but recall that **ClientUI** is still calling `endpointInstance.SendLocal(command)` which effectively sends the message to itself. But it doesn't have a handler for this message anymore.
```

At this point, we've managed to create two processes and achieve inter-process communication between them. Now let's try something different.

1. In Visual Studio's **Debug** menu, select **Detach All** so that we can close one console window without Visual Studio closing all the other windows as well. Alternatively, you can run the solution using **Debug** > **Start Without Debugging** or <kbd>Ctrl</kbd> + <kbd>F5</kbd>.
1. Close the **Sales** endpoint window so that only **ClientUI** is running.

> 1. Press <kbd>P</kbd> several times to send several messages to the **Sales** endpoint. Note that it works just fine; messages are sent, and nothing fails because the **Sales** endpoint happens to be offline.

Reviewer: ```suggestion
1. Press <kbd>P</kbd> several times to send several messages to the **Sales** endpoint. Note that it works just fine; messages are sent, and nothing fails even though the **Sales** endpoint is offline.
```
